Add support for CWL to Tooltester

fhembroff opened the pull request:
**Description**

This PR adds support for CWL workflows to the tooltester repository. It also adds two test cases that are CWL workflows, those being,
https://github.com/dockstore/dockstore-support/blob/f2b42624bff005f7a2c11aceca4579ed4e688901/tooltester/src/main/java/io/dockstore/tooltester/runWorkflow/WorkflowList.java#L53-L54

In order to use the tooltester `run-workflows` command, the user is required to specify the paths to two different dockstore config files, one that references an endpoint that runs `WDL` (only tested with Cromwell) and one that references that an endpoint that runs `CWL` (only tested with TOIL).

I found it extremely challenging to find workflows on qa.dockstore.org that have CWL workflows that can be used for testing, so I copied (and slightly modifed) two workflows from [here](https://github.com/aws/amazon-genomics-cli/tree/main/examples/demo-cwl-project/workflows). The workflows that I copied live in this repo (this repo is owned by myself, I can move it to one controlled by dockstore before the end of my term),
https://github.com/fhembroff/wes-testing

**Launching a workflow with TOIL**
I found that running workflows using TOIL is simpler than I thought (originally was looking at implementing a much more complicated approach), they are launched in a similar way to how you run workflows with WDL. The only differences are,

- You **must** use the `--inline-workflow` flag if you want to use any file that isn't the primary descriptor file (Like how `make-array.cwl` is used in this [workflow](https://qa.dockstore.org/workflows/github.com/fhembroff/wes-testing/manyJobs:main)). Currently, with cromwell the `--inline-workflow` will cause an error, and instead any file that is referenced that isn't the primary descriptor file must have an `s3` address.
- The test parameter file must follow the `--json` flag. There is no need (and if you include it, it will cause errors) for an `agcWrapper.json` file.
- You can reference https files in a workflows that is being run by TOIL in the agc.

For example, to launch this [workflow](https://qa.dockstore.org/workflows/github.com/fhembroff/wes-testing/manyJobs:main) with TOIL through the dockstore CLI you would use this command,
```
dockstore workflow wes launch --entry github.com/fhembroff/wes-testing/manyJobs:main --json Dockstore.json --inline-workflow
```
Where,
```
> cat Dockstore.json
{
    "script_file": {"class": "File", "path": "s3://flynn-test-bucket/make-data.js"},
    "script_argument_count": 10
}
```

One rather annoying thing about the TOIL endpoint is that the agc logs do not give you any information about what caused an error (all you get is `EXECUTOR_ERROR`), so debugging the TOIL endpoint can be very challenging. This is different to cromwell, which gives you a detailed log indicating what the errors were.

**How TOIL results are displayed**
Unlike when a workflow is run by Cromwell, TOIL workflows only provide timing information about the whole job (and not the individual tasks that run). 

**ISSUE**
[SEAB-5311](https://ucsc-cgl.atlassian.net/browse/SEAB-5311)

[SEAB-5311]: https://ucsc-cgl.atlassian.net/browse/SEAB-5311?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

fhembroff commented on tooltester/src/main/java/io/dockstore/tooltester/runWorkflow/WorkflowRunner.java:
```diff
@@ -163,13 +208,35 @@ public Boolean isWorkflowFinished() {
+                Date endTimeDate = null;
+                try {
+                    startTimeDate = parseISO8601Date(startTime + "Z");
+                    endTimeDate = parseISO8601Date(endTime + "Z");
```
The `Z` is added because the times that are given by the TOIL end point look like,
```
2023-03-20T16:49:23.664
```
(where the times are in UTC), the adding of the `Z` tells the parser that the timezone is UTC.

fhembroff replied:
The format given by the cromwell end point is,
```
2023-03-20T16:49:23.664Z
```

This is parsed perfectly by `parseISO8601Date` with no adjustments required.

denis-yuen commented:
> The workflows that I copied live in this repo (this repo is owned by myself, I can move it to one controlled by dockstore before the end of my term),
> https://github.com/fhembroff/wes-testing

Also create a ticket for this so we don't forget

fhembroff commented:
I have created two tickets,
1. [SEAB-5341](https://ucsc-cgl.atlassian.net/browse/SEAB-5341) this is to ensure that https://github.com/fhembroff/wes-testing is not used after the end of my co-op term.
2. https://github.com/dockstore/dockstore/issues/5435 to create documentation for launching CWL workflows through WES.

denis-yuen commented:
> One rather annoying thing about the TOIL endpoint is that the agc logs do not give you any information about what caused an error (all you get is `EXECUTOR_ERROR`), so debugging the TOIL endpoint can be very challenging. This is different to cromwell, which gives you a detailed log indicating what the errors were.

Interestingly, we do know some of the toil developers if this becomes a larger issue

denis-yuen commented on tooltester/src/main/java/io/dockstore/tooltester/runWorkflow/WorkflowList.java:
```diff
@@ -28,19 +29,29 @@ public class WorkflowList {
+                        new WorkflowRunner("github.com/gatk-workflows/seq-format-conversion/BAM-to-Unmapped-BAM", "3.0.0", "test-parameter-files/BAM-to-Unmapped-BAM-test-parameter-file.json", extendedGa4GhApi, workflowsApi, configFilePathWDL, configFilePathCWL),
+                        new WorkflowRunner("github.com/manning-lab/vcfToGds", "main", "test-parameter-files/vcfToGds-test-parameter-file.json", extendedGa4GhApi, workflowsApi, configFilePathWDL, configFilePathCWL),
+                        new WorkflowRunner("github.com/DataBiosphere/analysis_pipeline_WDL/assocation-aggregate-wdl", "v7.1.1", "test-parameter-files/assocation-aggregate-wdl-test-parameter-file.json", extendedGa4GhApi, workflowsApi, configFilePathWDL, configFilePathCWL),
+                        new WorkflowRunner("github.com/fhembroff/wes-testing/nontrivial", "main", "test-parameter-files/nontrivial-test-parameter-file.json", extendedGa4GhApi, workflowsApi, configFilePathWDL, configFilePathCWL),
```
Can also be done in that follow-up ticket, these should probably be tagged to be more resilient against future accidental changes

fhembroff replied:
I created this tag for my workflows `stable-version-for-testing-v1`.

denis-yuen commented on tooltester/src/main/java/io/dockstore/tooltester/runWorkflow/WorkflowRunner.java:
```diff
@@ -25,6 +25,9 @@
 import io.dockstore.openapi.client.model.Execution;
 import io.dockstore.openapi.client.model.FileWrapper;
+import io.dockstore.openapi.client.model.WorkflowSubClass;
+import io.swagger.client.api.WorkflowsApi;
```
this is mixing swagger and openapi (above) classes. 
See if you can use openapi only

fhembroff replied:
Fixed. We only use openapi imports. This caused a lot of changes.

I think it may've broke a small bit of code. However, I don't believe we were using that code anyway.

denis-yuen commented on tooltester/src/main/java/io/dockstore/tooltester/runWorkflow/WorkflowRunner.java:
```diff
@@ -90,10 +100,13 @@ public WorkflowRunner(String entry, String version, String pathOfTestParameter,
      */
-    public WorkflowRunner(String entry, String version, String relativePathToTestParameterFile, Ga4Ghv20Api ga4Ghv20Api, ExtendedGa4GhApi extendedGa4GhApi)  {
+    @SuppressWarnings("checkstyle:parameternumber")
+    public WorkflowRunner(String entry, String version, String relativePathToTestParameterFile, Ga4Ghv20Api ga4Ghv20Api, ExtendedGa4GhApi extendedGa4GhApi, WorkflowsApi workflowsApi, String WDLConfigFilePath, String CWLConfigFilePath)  {
```
the github code highlighting is a bit messed here, this isn't a typical naming pattern for parameters (leading caps)

denis-yuen replied:
also, this constructor is very similar to the first constructor, you can probably chain constructors to remove the repetitiveness https://www.scientecheasy.com/2020/06/java-constructor-chaining.html/

fhembroff replied:
Fixed

denis-yuen replied:
parameter naming is still non-standard, constructor is better though

fhembroff replied:
Sorry, completely missed the bit about the naming pattern.

It is now fixed.

denis-yuen reviewed (commented):
some changes to look into, overall approach is good

coverbeck commented:
> One rather annoying thing about the TOIL endpoint is that the agc logs do not give you any information about what caused an error (all you get is EXECUTOR_ERROR), so debugging the TOIL endpoint can be very challenging. This is different to cromwell, which gives you a detailed log indicating what the errors were.

@adamnovak FYI

coverbeck commented:
> You must use the --inline-workflow flag if you want to use any file that isn't the primary descriptor file (Like how make-array.cwl is used in this [workflow](https://qa.dockstore.org/my-workflows/github.com/fhembroff/wes-testing/manyJobs)). Currently, with cromwell the --inline-workflow will cause an error, and instead any file that is referenced that isn't the primary descriptor file must have an s3 address.
    The test parameter file must follow the --json flag. There is no need (and if you include it, it will cause errors) for an agcWrapper.json file.
    You can reference https files in a workflows that is being run by TOIL in the agc.

FYI, your links are to `my-workflows`, which we can't see (but I found the public versions). 

By `You must use the --inline-workflow flag if you want to use any file that isn't the primary descriptor file`, do you mean if there is more than one descriptor in the workflow, you need to use that switch? And if there is only one descriptor, presumably it works with or without the `--inline-workflow` switch? If so, maybe it's easier to just say in the doc, when you get to it :), to always use the `inline-workflow` switch when running on WES/Toil?

> For example, to launch this [workflow](https://qa.dockstore.org/my-workflows/github.com/fhembroff/wes-testing/manyJobs) with TOIL through the dockstore CLI you would use this command,

> dockstore workflow wes launch --entry github.com/fhembroff/wes-testing/manyJobs:main --json Dockstore.json

Shouldn't this this have `--inline-workflow`? I know it's just a PR description, but I'm trying to understand.

coverbeck reviewed (approved):
Nice!

fhembroff commented:
> FYI, your links are to `my-workflows`, which we can't see (but I found the public versions).

Oops, my bad. I have fixed the links.

> By You must use the --inline-workflow flag if you want to use any file that isn't the primary descriptor file, do you mean if there is more than one descriptor in the workflow, you need to use that switch? And if there is only one descriptor, presumably it works with or without the --inline-workflow switch? If so, maybe it's easier to just say in the doc, when you get to it :), to always use the inline-workflow switch when running on WES/Toil?

Yes, that is correct. Will do!

> Shouldn't this this have --inline-workflow? I know it's just a PR description, but I'm trying to understand.


That is correct, I have updated the description.

kathy-t commented on tooltester/src/test/java/io/dockstore/tooltester/helper/GA4GHHelperTest.java:
```diff
@@ -41,6 +42,7 @@ public void runnerSupportsDescriptorType() {
         assertFalse(GA4GHHelper.runnerSupportsDescriptorType("cromwell", ToolVersion.DescriptorTypeEnum.NFL));
     }
 
+    @Disabled
```
Same comment about disabled test

fhembroff replied:
Reason added

fhembroff commented on tooltester/src/main/java/io/dockstore/tooltester/helper/GA4GHHelper.java:
```diff
@@ -108,8 +111,13 @@ static List<Tool> filterTools(List<Tool> allTools, boolean verified, List<String
                                 .collect(Collectors.toList()));
+                // The above was originally
+                /*
+                tool.getVersions().parallelStream().filter(p -> matchVerifiedSource(verifiedSources, p.getVerifiedSource()))
```
I have added the below comment,
```
TODO: verify that the code still works as intended with the above change
```

fhembroff commented on tooltester/src/main/java/io/dockstore/tooltester/runWorkflow/WorkflowRunner.java:
```diff
@@ -116,6 +125,24 @@ public WorkflowRunner(String entry, String version, String relativePathToTestPar
+                break;
+
+            default:
+                errorMessage("The descriptor type of " + workflow.getWorkflowName() + "(which is " + descriptorType.toString() + ") is not supported", GENERIC_ERROR);
```
Good catch, I have fixed this error, thanks for spotting it.

fhembroff commented on tooltester/src/test/java/io/dockstore/tooltester/helper/DockstoreEntryHelperTest.java:
```diff
@@ -38,11 +39,12 @@ public class DockstoreEntryHelperTest {
         containersApi = new ContainersApi(defaultApiClient);
     }
 
+    @Disabled
```
Reason added

david4096 reviewed (approved):
Really pleased to see tooltester being brought up to date. Also really glad to hear about your experience with toil. We definitely should be able to get our feature requests handled on that one and I generally want to encourage that collaboration.
